tool/src/c2/formatter.rs

```diff
@@ -0,0 +1,107 @@
+        self.tcx.resolve_type(id).name().as_str().into()
+    }
+    /// Resolve and format the name of a type for use in header names
+    pub fn fmt_header_name(&self, id: TypeId) -> Cow<'tcx, str> {
+        self.fmt_type_name(id)
+    }
+    /// Format an enum variant.
+    pub fn fmt_enum_variant(&self, variant: &'tcx hir::EnumVariant) -> Cow<'tcx, str> {
+        variant.name.as_str().into()
+    }
+    /// Format a field name or parameter name
+    // might need splitting in the future if we decide to support renames here
+    pub fn fmt_param_name<'a>(&self, ident: &'a str) -> Cow<'a, str> {
+        ident.into()
+    }
+
+    /// Format a method
+    pub fn fmt_method_name(&self, ty: TypeId, method: &hir::Method) -> String {
+        let ty_name = self.fmt_type_name(ty);
+        let method_name = method.name.as_str();
+        format!("{ty_name}_{method_name}")
+    }
+
+    /// Given a mutability, format a `const ` prefix for pointers if necessary,
+    /// including a space for prepending
+    pub fn fmt_constness(&self, mutability: hir::Mutability) -> &str {
+        if mutability.is_mutable() {
+            ""
+        } else {
+            "const "
+        }
+    }
+
+    /// Generates an identifier that uniquely identifies the given *C* type.
+    /// Rust types that map to the same C type will get the same C identifier
+    /// (e.g. &mut Foo and Option<&mut Foo> are all the same)
+    ///
+    /// This is primarily used for generating structs for result types,
+    /// which require one struct for each distinct instance.
+    pub fn fmt_type_name_uniquely<P: hir::TyPosition>(&self, ty: &'tcx Type<P>) -> Cow<'tcx, str> {
+        match ty {
+            Type::Primitive(p) => self.fmt_primitive_as_c(*p),
+            Type::Opaque(o) => {
+                let o_name = self.fmt_type_name(o.tcx_id.into());
+                // Todo (breaking): box should be unified with the mutable branch
+                let ownership = match o.owner.mutability() {
+                    None => "box_",
+                    Some(hir::Mutability::Mutable) => "",
+                    Some(hir::Mutability::Immutable) => "const_",
+                };
+                format!("{ownership}{o_name}").into()
+            }
+            Type::Struct(s) => self.fmt_type_name(P::id_for_path(s)),
+            Type::Enum(e) => self.fmt_type_name(e.tcx_id.into()),
+            Type::Slice(hir::Slice::Str(_)) => "str_ref".into(),
+            Type::Slice(hir::Slice::Primitive(borrow, p)) => {
+                let constness = if borrow.mutability.is_mutable() {
+                    ""
+                } else {
+                    "const_"
+                };
+                let prim = self.fmt_primitive_as_c(*p);
+                format!("ref_{constness}prim_slice_{prim}").into()
+            }
+        }
+    }
+
+    /// Get the primitive type as a C type
+    pub fn fmt_primitive_as_c(&self, prim: hir::PrimitiveType) -> Cow<'static, str> {
```

Comment:
Just consistency, really. Could be either.

We could use invariants but IMO this would involve a lot of conversions for little gain: we're producing these, we shouldn't need to check them ourselves. I find strck more useful for user generated stuff.